database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -85,8 +85,32 @@ public String parseSnapshot(DataSnapshot snapshot) {
                 return snapshot.getKey();
             }
         });
+    }
+
+    @Override
+    public ChangeEventListener addChangeEventListener(@NonNull ChangeEventListener listener) {
+        boolean wasListening = isListening();
+        super.addChangeEventListener(listener);
+
+        // Only start listening when the first listener is added
+        if (!wasListening) {
+            mKeySnapshots.addChangeEventListener(this);
+        }
 
-        mKeySnapshots.addChangeEventListener(this);
+        return listener;
+    }
+
+    @Override
+    public void removeChangeEventListener(@NonNull ChangeEventListener listener) {
+        super.removeChangeEventListener(listener);
+        if (!isListening()) {
+            mKeySnapshots.removeChangeEventListener(this);
```

Comment:
I can't believe I never noticed this, but we never actually removed the `FirebaseArray` listener! 😱